pydantic_ai_slim/pydantic_ai/agent/__init__.py

```diff
@@ -592,20 +600,27 @@ async def main():
         usage_limits = usage_limits or _usage.UsageLimits()
 
         async def get_instructions(run_context: RunContext[AgentDepsT]) -> str | None:
-            parts = [
-                self._instructions,
-                *[await func.run(run_context) for func in self._instructions_functions],
-            ]
+            parts: list[str | None] = []
+            # If instructions override is set, ignore registered instruction functions and base instructions
+            if override_instructions := self._override_instructions.get():
+                base_text = override_instructions.value
+                parts = [base_text]
+            else:
+                parts = [
+                    self._instructions,
+                    *[await func.run(run_context) for func in self._instructions_functions],
+                ]
 
             model_profile = model_used.profile
             if isinstance(output_schema, _output.PromptedOutputSchema):
                 instructions = output_schema.instructions(model_profile.prompted_output_template)
                 parts.append(instructions)
 
             parts = [p for p in parts if p]
-            if not parts:
+            parts_to_format = [p for p in parts if p is not None]
+            if not parts_to_format:
                 return None
-            return '\n\n'.join(parts).strip()
+            return '\n\n'.join(parts_to_format).strip()
```

Comment:
The existing logic filters out empty strings as well, why do we now only filter `None`s? Note also that if we do want this change, we can just change `parts` instead of having both `parts` and `parts_to_format` :)